server/clientRoutes.js

```diff
@@ -0,0 +1,124 @@
+  router.get('/default', (req, res) => {
+    if (!clients.has(defaultId))
+      return res.status(500).json({
+        error: {
+          message: `Error retrieving default client: ${defaultId}`,
+          code: 500
+        }
+      });
+
+    const defaultClient = {
+      id: defaultId,
+      ...clientInfo[defaultId]
+    };
+    res.status(200).json(defaultClient);
+  });
+
+  // middleware for setting constants based on
+  // the route being used
+  router.use('/:id', (req, res, next) => {
+    id = req.params.id;
+    if (!clients.has(id))
+      return res.status(404).json({
+        error: {
+          message: `Sorry, there was no client with the id ${id}`,
+          code: 404
+        }
+      });
+    const clientObj = clients.get(id);
+    config = clientObj.config;
+    assert(config instanceof Config, 'client needs bcfg config');
+
+    // store clients to be used in the routes
+    // handled in the next middleware
+    reqClients = {
+      node: clientObj.nodeClient,
+      wallet: clientObj.walletClient,
+      multisig: clientObj.multisigClient
+    };
+    next();
+  });
+
+  // all routes for a given client
+  // the client param dictates client, node,
+  // wallet, or multisig, will be used
+  router.use('/:id/:client/', async (req, res) => {
+    const { method, path, body, query, params } = req;
+    const client = reqClients[params.client];
+    if (!client)
+      return res.status(404).json({
+        error: {
+          message: `Requested client ${params.client} for ${id} does not exist`,
+          code: 404
+        }
+      });
+
+    // use query params for GET request, otherwise use body
+    const payload = method === 'GET' ? query : body;
+    try {
+      logger.debug(
+        `client: ${client.constructor.name}, method: ${method},`,
+        `path: ${path}`
+      );
+      logger.debug('query:', query);
+      logger.debug('body:', body);
+
+      // proxy the token
+      token = client.token;
+      if ('token' in payload) {
+        client.token = payload.token;
+        logger.debug('Using custom client token');
+      }
+      const response = await client.request(method, path, payload);
+      logger.debug('server response:', response ? response : 'null');
+      if (response) return res.status(200).json(response);
+      // return 404 when response is null due to
+      // resource not being found on server
+      return res.status(404).json({ error: { message: 'resource not found' } });
+    } catch (e) {
+      logger.error(`Error querying ${client.constructor.name}:`, e);
+      return res
+        .status(502)
+        .send({ error: { message: e.message, code: e.code, type: e.type } });
+    } finally {
+      // always reassign the original token
+      client.token = token;
+    }
+  });
+
+  // get info about a specific client
+  router.get('/:id', (req, res) => {
+    const info = {
+      network: config.str('network'),
```

Comment:
If you change the network, will this contain the updated values? It looks like to update the network, you would have to invoke `clientsRouter` with a new config. Is that the intended way of using this?

Reply:
There is no mechanism to change the network server side. There is a default client now [that is set](https://github.com/bpanel-org/bpanel/pull/79/files#diff-c1969c591890910e4499a3c110fbbcdbR24), but this is primarily to be backwards compatible and support he legacy setup. Instead what we'll do to support multiple networks is that clients need to be explicit about which node they want to query based off of this `:id` param. The `config` variable gets set based off of that param [in this middleware](https://github.com/bpanel-org/bpanel/pull/79/files#diff-c1969c591890910e4499a3c110fbbcdbL27) and is used for subsequent routes. So all the front end has to do is query `[HOST]/clients/:id/:type/<bclient/hs-client request>`. This is made even easier if you're using the new BPClient instance.

Reply:
Ok, cool, looks good to me